### PR TITLE
feat(ensemble): add ExtraSurvivalTrees

### LINKED ISSUE #194: feat(ensemble): support ExtraSurvivalTrees
**Is your feature request related to a problem? Please describe.**
This feature request is releated to `ExtraSurvivalTrees`, a natural extension of `RandomSurvivalForest` that introduces more randomness into the building process of the forest.

`ExtraSurvivalTrees` is supported in [pysurvival](https://square.github.io/pysurvival/models/extra_survival_trees.html). However, that package is no longer maintained any more, and I think it would be nice to implement the same model in sksurv.

**Describe the solution you'd like**
This should be easy since survival trees in sksurv already support the `random` splitter (i.e., the extra tree). We could add the implementation of `ExtraSurvivalTrees` in `forest.py`, with its `base_estimator` attribute set to `SurvivalTree(splitter='random')`.

``` python
class ExtraSurvivalTrees(RandomSurvivalForest):

    def __init__(self,
                 n_estimators=100,
                 max_depth=None,
                 min_samples_split=6,
                 min_samples_leaf=3,
                 min_weight_fraction_leaf=0.,
                 max_features="auto",
                 max_leaf_nodes=None,
                 bootstrap=True,
                 oob_score=False,
                 n_jobs=None,
                 random_state=None,
                 verbose=0,
                 warm_start=False,
                 max_samples=None):
        super().__init__(
            base_estimator=SurvivalTree(splitter="random"),
            n_estimators=n_estimators,
            estimator_params=("max_depth",
                              "min_samples_split",
                              "min_samples_leaf",
                              "min_weight_fraction_leaf",
                              "max_features",
                              "max_leaf_nodes",
                              "random_state"),
            bootstrap=bootstrap,
            oob_score=oob_score,
            n_jobs=n_jobs,
            random_state=random_state,
            verbose=verbose,
            warm_start=warm_start,
            max_samples=max_samples)

        self.max_depth = max_depth
        self.min_samples_split = min_samples_split
        self.min_samples_leaf = min_samples_leaf
        self.min_weight_fraction_leaf = min_weight_fraction_leaf
        self.max_features = max_features
        self.max_leaf_nodes = max_leaf_nodes
```

### PR DESCRIPTION
**Pull request by xuyxu**
**Checklist**

- [x] closes #194
- [x] py.test passes
- [x] tests are included
- [x] code is well formatted
- [x] documentation renders correctly

**What does this implement/fix? Explain your changes**

This PR adds the implementation of `ExtraSurvivalTrees`. To achieve this,

* An abstract class `_BaseSurvivalForest` is added, following the code organization of `ForestClassifier`, `RandomForestClassifier`, and `ExtraTreesClassifier` in Scikit-Learn
* Refactor `RandomSurvivalForest` and `ExtraSurvivalTrees` on top of `_BaseSurvivalForest`

### CODE REVIEW

File: tests/test_forest.py [hunk 1]
```diff
@@ -107,11 +116,12 @@ def test_oob_score(make_whas500):
     assert round(abs(forest.oob_score_ - 0.753010685), 6) == 0.0
```
Comment (xuyxu) [inline tests/test_forest.py:116]: I noticed that two unit tests `test_fit_predict` and `test_oob_score` directly put the expected results (for RandomSurvivalForest) in the assertion statement, and not sure how to modify it to adapt to `ExtraSurvivalTrees`. Will appreciate your suggestions very much ;-)

Comment (sebp) [reply tests/test_forest.py:116]: You can pass in the expected value via pytest's parametrize too.

**Review by sebp — CHANGES_REQUESTED** [review]
Looks good for the most part, only some small changes are required.

**Comment by xuyxu** [comment]
Hi @sebp, thanks for your suggestions. I have made some modifications to improve the code quality.

**Comment by sebp** [comment]
@xuyxu Thanks for the update. Can you please add `ExtraSurvivalTrees` to doc/api/ensemble.rst too?

**Comment by sebp** [comment]
Merged in commit 6f9c09763a4e386928fd1e2b5654f0a6a87bd94e

Thanks again!

**Pull request closed without merging**